pkg/relayer/session/persistence.go

```diff
@@ -0,0 +1,200 @@
+		// 5. Past proof window: Delete as the session is completely expired
+
+		claimWindowCloseHeight := sharedtypes.GetClaimWindowCloseHeight(sharedParams, sessionEndHeight)
+		proofWindowCloseHeight := sharedtypes.GetProofWindowCloseHeight(sharedParams, sessionEndHeight)
+
+		// Check if session is expired (past proof submission window)
+		// These sessions can no longer reach settlement and should be deleted
+		if height > proofWindowCloseHeight {
+			if storeErr := rs.deletePersistedSessionTree(persistedSMT); storeErr != nil {
+				sessionLogger.Error().Err(storeErr).Msg("failed to delete outdated session tree, skipping")
+				continue
+			}
+
+			sessionLogger.Info().
+				Int64("proof_window_close", proofWindowCloseHeight).
+				Int64("current_height", height).
+				Msg("Session deleted: expired past proof submission window")
+
+			continue
+		}
+
+		// For sessions past the claim window, check if a claim exists onchain
+		// If no claim exists, the session can't progress to settlement and should be deleted
+		var claim *prooftypes.Claim
+		if height > claimWindowCloseHeight {
+			claim, queryErr := rs.proofQueryClient.GetClaim(ctx, supplierOperatorAddress, sessionId)
+			// No claim was created onchain, delete the session from the store.
+			if queryErr != nil || claim == nil {
+				if storeErr := rs.deletePersistedSessionTree(persistedSMT); storeErr != nil {
+					sessionLogger.Error().Err(storeErr).Msg("failed to delete outdated session without claim, skipping")
+					continue
+				}
+
+				sessionLogger.Info().Msg("deleting session has no claim created onchain, deleting session metadata")
+				continue
+			}
+		}
+
+		// The session has still a chance to reach settlement.
+		// Add it to the sessionTrees map of blockHeight->sessionId->supplierOperatorAddress->sessionTree
+		sessionTreesWithEndHeight, ok := rs.sessionsTrees[sessionEndHeight]
+
+		// If there is no map for sessions at the sessionEndHeight, create one.
+		if !ok {
+			sessionTreesWithEndHeight = make(map[string]map[string]relayer.SessionTree)
+			rs.sessionsTrees[sessionEndHeight] = sessionTreesWithEndHeight
+		}
+
+		// Get the sessionTreeWithSessionId for the given session.
+		sessionTreeWithSessionId, ok := sessionTreesWithEndHeight[sessionId]
+
+		// If there is no map for session trees with the session id, create one.
+		if !ok {
+			sessionTreeWithSessionId = make(map[string]relayer.SessionTree)
+			sessionTreesWithEndHeight[sessionId] = sessionTreeWithSessionId
+		}
+
+		_, ok = sessionTreeWithSessionId[supplierOperatorAddress]
+		if ok {
+			sessionLogger.Error().Msg("the session tree already exists, skipping")
+			continue
+		}
+
+		sessionTree, treeErr := ImportSessionTree(persistedSMT, claim, rs.storesDirectory, rs.logger)
+		if treeErr != nil {
+			sessionLogger.Error().Err(treeErr).Msg("failed to import session tree")
+			continue
+		}
+
+		sessionTreeWithSessionId[supplierOperatorAddress] = sessionTree
```

Comment:
#PUC (if I'm understanding correctly) that this is where each loaded persisted session tree is actually included in the set of session trees which are processed by the observable pipeline.